### ISSUE TITLE
[Dregora - Issue & Fix?] Sapphire City Generation Incomplete

### ISSUE THREAD

**Issue by TheOneKevin** [NLBlackEagle/Dregora#22]
Environment: Spigot 1.12.2 (Implementing API version 1.12.2-R0.1-SNAPSHOT)
Dregora Version: Dregora-1.12.2-v1.5.1+HotFix+(+Generation+Lag+)
Plugins: OpenTerrainGenerator-1.12.2+-+v6
Mode: FromImage
Issue: Sapphire City does not generate properly. Only 1 chunk generates.
Steps to Reproduce: Configure server from latest Dregora configs, and start the server as normal.

Upon closer inspection, the file `WorldObjects\Structures\Sap53.bo3` has a typo at lines 3481-3488 as:
```
Branch(0,0,-16,,sap54,NORTH,100.0,0)
Branch(16,0,-16,,sap63,NORTH,100.0,0)
Branch(-16,0,-16,,sap41,NORTH,100.0,0)
Branch(-16,0,0,,sap42,NORTH,100.0,0)
Branch(16,0,0,,sap64,NORTH,100.0,0)
Branch(0,0,16,,sap52,NORTH,100.0,0)
Branch(16,0,16,,sap65,NORTH,100.0,0)
Branch(-16,0,16,,sap43,NORTH,100.0,0)
```
Should really be
```
Branch(0,0,-16,sap54,NORTH,100.0,0)
Branch(16,0,-16,sap63,NORTH,100.0,0)
Branch(-16,0,-16,sap41,NORTH,100.0,0)
Branch(-16,0,0,sap42,NORTH,100.0,0)
Branch(16,0,0,sap64,NORTH,100.0,0)
Branch(0,0,16,sap52,NORTH,100.0,0)
Branch(16,0,16,sap65,NORTH,100.0,0)
Branch(-16,0,16,sap43,NORTH,100.0,0)
```
The entire city generated successfully after modifications. Could this be perhaps related to #21?

However, something else was not quite right as the city quickly burned the ground a few minutes after spawning.
![2019-04-07_15 37 06](https://user-images.githubusercontent.com/16702203/55688841-7d3bc880-594b-11e9-8de6-38068898831a.png)
`doFireTick` had to be disabled.

**Comment by TheOneKevin** [comment]
So, after some close inspection, I ran grep on all the bo3 files to search for the double comma error, and there were a lot. `grep -oP "Branch\(-?\d+,-?\d+,-?\d+,{1}" -rnw "WorldObjects/" | wc -l` returned 4392. There are 4392 files that have this bug. I have attached a list of them all.
[out.txt](https://github.com/NLBlackEagle/Dregora/files/3052018/out.txt)

Thus, simply running `sed -E "s/(Branch\(-?[0-9]+,-?[0-9]+,-?[0-9]+),{2}/\1,/g"` on affected files should fix the problem. In this case, `find WorldObjects/ -type f -name "*.bo3" -exec sed -i -E "s/(Branch\(-?[0-9]+,-?[0-9]+,-?[0-9]+),{2}/\1,/g" {} \;` did the trick. I just fired up WSL (I'm on Windows), cd'ed into the world_dregora directory and ran it. Then I ran grep later to confirm that indeed the double commas were gone.

I still have no idea whether or not the double commas were intentional or not, and whether or not it contributed to many generation errors. I know that it definitely fixed my Sapphire City issue.

**Comment by TheOneKevin** [comment]
Interestingly enough, Notepad++ does offer the ability to find & replace in directories.
I've managed to fix RF_Village generation errors by recursively applying the following regex: `(WeightedBranch\(.*\w+,[A-Z]+,\d+.\d+)(,0)` where the replacement regex is `\1`
Running this repeatedly until all occurrences are gone fixes it (it took over 20 replacements and a long time).

I hope OP can fix this in the upcoming release instead of employing hacky regex fixes.

![2019-04-08_21 03 29](https://user-images.githubusercontent.com/16702203/55766242-ca01cb00-5a41-11e9-96ff-053bcd62608c.png)
Beautiful RF_Village finally generates!

**Comment by TheOneKevin** [comment]
Ok, I'm not sure why there are so many errors (perhaps I'm not using the right OTG version)?
All instances of `RandomB(` must be replaced with `RandomBlock(` or your server will crash. Also, all instances of `SPRUCE_WOOD_STAIRS:0:0` must be replaced with `SPRUCE_WOOD_STAIRS:0` or your server will crash too.

**Comment by alxcool** [comment]
Hi! I've experimented the same issue as...
1.12.2 old launcher, forge2768 single player
OTG + Dregora + Biome Bundle in the mods folder...
Not sure if it was the previous one ... Dregora-1.12.2-v1.5+Stable

What I would like to say is, that I manage to use a backup file in porder to extinguis town's fire while on doFireTick = false. In order to learn what cause it. In fact, the town own 12 fireplace where 11 of them are placed way too close to wood materia or wool, or .... An another point anoying for a firefighter point of view is... Where is the door leading to that fireplace? As most town building are divided like appartments building with many doors. About the room who contain an active fireplace, these are not properly stacked together. I mean It could be easier to plan a building with fireproof material when your netherrack are only seperated by height. Right now, Its a maze just to figure out where the fire noise came from and where is the next one, as they are not bellow or on top of the previous one.

I would Like to offer my help to fix the city, by moving some blocks arround, then by saving the building in a bo3 format... I just never did such a thing so... wish me luck!

Note... I don't know how I was able to find that many loot chest filled with loots but thk anyway ... It was quite fun to look for such a thing! Note sincce I've changed my laptop using the new minecraft launcher, it look like if many of the structure problem occur more often and the loots and chests are both gone :( unfortunatly.

**Comment by NLBlackEagle** [comment]
- Issue of sapphire city not spawning are resolved as of 1.62 ( This issue may have been fixed on an earlier build )
- Issues regarding RF villages are also fixed as of 1.62
- Fire issues have been fixed as of 1.62 as well. ( I might have missed some instances though, added big sweep with tirkrate 1000 to to-do list for 1.63  )

Apologies for my absense on github, I used to have a discord webhook which apperantly stopped working. Will fix this asap to keep on track with issues.
